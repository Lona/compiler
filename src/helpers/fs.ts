import * as fs from 'fs'
import * as path from 'path'

export type FSWrapper = {
  readFile(filePath: string): Promise<string>
  writeFile(filePath: string, data: string): Promise<void>
  copyDir(dirPath: string, output?: string): Promise<void>
}

export const createFSWrapper = (
  workspacePath: string,
  _outputPath?: unknown
): FSWrapper => {
  const outputPath =
    typeof _outputPath === 'string'
      ? _outputPath
      : path.join(process.cwd(), 'lona-generated')
  const fsWrapper = {
    readFile(filePath: string) {
      return fs.promises.readFile(
        path.resolve(workspacePath, filePath),
        'utf-8'
      )
    },
    writeFile(filePath: string, data: string) {
      const resolvedPath = path.resolve(outputPath, filePath)
      fs.mkdirSync(path.dirname(resolvedPath), { recursive: true })
      return fs.promises.writeFile(resolvedPath, data, 'utf-8')
    },
    async copyDir(dirPath: string, output: string = '.') {
      const resolvedPath = path.resolve(workspacePath, dirPath)
      const files = await fs.promises.readdir(resolvedPath)

      await Promise.all(
        files.map(async x => {
          if (
            (await fs.promises.stat(path.join(resolvedPath, x))).isDirectory()
          ) {
            return
          }

          return fsWrapper.writeFile(
            path.join(output, x),
            await fsWrapper.readFile(path.join(dirPath, x))
          )
        })
      )
    },
  }

  return fsWrapper
}
