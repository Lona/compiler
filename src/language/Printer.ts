import { doc, Doc } from 'prettier'
import { Definition, Value } from './Parser'
import { StateDefinition, Token, PrintPattern } from './Lexer'

const {
  builders: { concat },
} = doc

export function formatToken(
  lexerDefinition: StateDefinition[],
  token: Token
): Doc {
  const rules = lexerDefinition.flatMap(state => state.rules)

  const rule = rules.find(rule => rule.name === token.type)

  if (!rule) {
    throw new Error(`Couldn't find rule for token: ${token.type}`)
  }

  function formatPrintPattern(printPattern: PrintPattern): Doc {
    switch (printPattern.type) {
      case 'literal': {
        return printPattern.value
      }
      case 'indexReference': {
        return token.groups[printPattern.value]
      }
      case 'sequence': {
        return concat(printPattern.value.map(formatPrintPattern))
      }
    }
  }

  return rule.print ? formatPrintPattern(rule.print) : ''
}

export function formatTokens(
  lexerDefinition: StateDefinition[],
  tokens: Token[]
): Doc {
  return concat(tokens.map(formatToken.bind(null, lexerDefinition)))
}

export function print(
  document: Doc,
  options: doc.printer.Options = {
    printWidth: 80,
    tabWidth: 2,
    useTabs: false,
  }
) {
  return doc.printer.printDocToString(document, options).formatted
}
