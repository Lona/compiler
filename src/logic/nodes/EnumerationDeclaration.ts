import { LogicAST as AST } from '@lona/serialization'
import { compact } from '../../utils/sequence'
import { EvaluationVisitor } from '../evaluationVisitor'
import NamespaceVisitor from '../namespaceVisitor'
import { DefaultArguments, RecordMemory } from '../runtime/memory'
import { ScopeVisitor } from '../scopeVisitor'
import { StaticType } from '../staticType'
import { TypeCheckerVisitor } from '../typeChecker'
import { substitute } from '../typeUnifier'
import { IDeclaration, Node } from './interfaces'

export class EnumerationDeclaration extends Node<AST.EnumerationDeclaration>
  implements IDeclaration {
  namespaceEnter(visitor: NamespaceVisitor): void {
    const {
      name: { name, id },
    } = this.syntaxNode.data

    visitor.declareType(name, id)
    visitor.pushPathComponent(name)
  }

  namespaceLeave(visitor: NamespaceVisitor): void {
    const { cases } = this.syntaxNode.data

    // Add initializers for each case into the namespace
    cases.forEach(enumCase => {
      switch (enumCase.type) {
        case 'placeholder':
          break
        case 'enumerationCase':
          visitor.declareValue(enumCase.data.name.name, enumCase.data.name.id)
      }
    })

    visitor.popPathComponent()
  }

  scopeEnter(visitor: ScopeVisitor): void {
    const { name, genericParameters } = this.syntaxNode.data

    visitor.addTypeToScope(name)

    visitor.pushNamespace(name.name)

    genericParameters.forEach(parameter => {
      switch (parameter.type) {
        case 'parameter':
          visitor.addTypeToScope(parameter.data.name)
        case 'placeholder':
          break
      }
    })
  }

  scopeLeave(visitor: ScopeVisitor): void {
    visitor.popNamespace()
  }

  typeCheckerEnter(visitor: TypeCheckerVisitor): void {
    const { genericParameters, cases, name } = this.syntaxNode.data
    const { typeChecker } = visitor

    const genericNames = compact(
      genericParameters.map(param =>
        param.type === 'parameter' ? param.data.name.name : undefined
      )
    )

    const genericsInScope: [string, string][] = genericNames.map(x => [
      x,
      typeChecker.typeNameGenerator.next(),
    ])

    const universalTypes = genericNames.map<StaticType>((x, i) => ({
      type: 'generic',
      name: genericsInScope[i][1],
    }))

    const returnType: StaticType = {
      type: 'constant',
      name: name.name,
      parameters: universalTypes,
    }

    cases.forEach(enumCase => {
      if (enumCase.type === 'placeholder') return

      const parameterTypes = compact(
        enumCase.data.associatedValueTypes.map(annotation => {
          if (annotation.type === 'placeholder') return

          return {
            label: undefined,
            type: visitor.unificationType(
              genericsInScope,
              () => typeChecker.typeNameGenerator.next(),
              annotation
            ),
          }
        })
      )

      const functionType: StaticType = {
        type: 'function',
        returnType,
        arguments: parameterTypes,
      }

      typeChecker.nodes[enumCase.data.name.id] = functionType
      typeChecker.patternTypes[enumCase.data.name.id] = functionType
    })

    /* Not used for unification, but used for convenience in evaluation */
    typeChecker.nodes[name.id] = returnType
    typeChecker.patternTypes[name.id] = returnType
  }

  typeCheckerLeave(visitor: TypeCheckerVisitor): void {}

  evaluationEnter(visitor: EvaluationVisitor) {
    const { name, cases } = this.syntaxNode.data
    const { typeChecker, substitution, reporter } = visitor

    const type = typeChecker.patternTypes[name.id]

    if (!type) {
      reporter.error('unknown enumberation type')
      return
    }

    const enumType = substitute(substitution, type)

    cases.forEach(enumCase => {
      if (enumCase.type !== 'enumerationCase') return

      const { name: caseName } = enumCase.data

      const caseType = substitute(
        substitution,
        typeChecker.nodes[enumCase.data.name.id]
      )

      if (caseType.type !== 'function') {
        throw new Error('Enum case type must be a function')
      }

      const defaultArguments: DefaultArguments = Object.fromEntries(
        caseType.arguments.map(({ label, type }, index) => {
          return [
            typeof label === 'string' ? label : index.toString(),
            [substitute(substitution, type), undefined],
          ]
        })
      )

      visitor.addValue(caseName.id, {
        type: enumType,
        memory: {
          type: 'function',
          value: {
            f: (record: RecordMemory) => {
              return {
                type: 'enum',
                value: caseName.name,
                data: Object.values(record),
              }
            },
            defaultArguments,
          },
        },
      })
    })
  }
}
