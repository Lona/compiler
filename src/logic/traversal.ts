import { LogicAST as AST } from '@lona/serialization'
import { withOptions, IndexPath } from 'tree-visit'

const { visit, find, findAll } = withOptions({
  getChildren: AST.subNodes,
})

// Overload to support type guard
export function findNode<
  A extends AST.SyntaxNode['type'],
  T extends AST.SyntaxNode & { type: A }
>(
  rootNode: Parameters<typeof find>[0],
  predicate: (node: AST.SyntaxNode, indexPath: IndexPath) => node is T
): T | undefined

export function findNode(
  node: Parameters<typeof find>[0],
  predicate: (node: AST.SyntaxNode, indexPath: IndexPath) => boolean
): ReturnType<typeof find> {
  return find(node, predicate)
}

export const findNodes = findAll

export { visit }
