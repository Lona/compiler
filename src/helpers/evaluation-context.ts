import * as fs from 'fs'
import * as path from 'path'
import { Config } from '../utils/config'
import * as LogicAST from './logic-ast'
import * as LogicScope from './logic-scope'
import * as LogicUnify from './logic-unify'
import * as LogicEvaluate from './logic-evaluate'
import { Reporter } from './reporter'
import { uuid, nonNullable } from '../utils'

/**
 * Iterates through the import declaration of the program
 * and "resolve them", eg. returns a program that contains both
 * the content of imported files and the current program
 */
function resolveImports(
  program: LogicAST.AST.Program,
  reporter: Reporter,
  existingImports: string[] = []
): LogicAST.AST.Program {
  return {
    type: 'program',
    data: {
      id: uuid(),
      block: program.data.block
        .map(x => {
          if (x.type !== 'declaration') {
            return [x]
          }
          if (x.data.content.type !== 'importDeclaration') {
            return [x]
          }

          const libraryName = x.data.content.data.name.name

          if (existingImports.indexOf(libraryName) !== -1) {
            return [x]
          }

          const libraryPath = path.join(
            __dirname,
            '../../static/logic',
            `${libraryName}.logic`
          )
          const libraryExists = fs.existsSync(libraryPath)

          if (!libraryExists) {
            reporter.error(
              `Failed to find library ${libraryName} at path ${libraryPath}`
            )
            return [x]
          }

          const library = LogicAST.makeProgram(
            JSON.parse(fs.readFileSync(libraryPath, 'utf8'))
          )

          if (!library) {
            reporter.error(`Failed to import library ${libraryName}`)
            return [x]
          }

          const resolvedLibrary = resolveImports(
            library,
            reporter,
            existingImports.concat(libraryName)
          )

          return [x, ...resolvedLibrary.data.block]
        })
        .reduce((prev, x) => prev.concat(x), []),
    },
  }
}

export const generate = (config: Config, reporter: Reporter) => {
  const standardLibsPath = path.join(__dirname, '../../static/logic')
  const standardLibs = fs.readdirSync(standardLibsPath)

  const libraryFiles: LogicAST.AST.Program[] = standardLibs.map(
    x =>
      LogicAST.makeProgram(
        JSON.parse(fs.readFileSync(path.join(standardLibsPath, x), 'utf8'))
      ) as LogicAST.AST.Program
  )

  const standardLibsProgram = LogicAST.joinPrograms(libraryFiles)

  const logicPrograms = Object.keys(config.logicFiles)
    .map(k => {
      const node = LogicAST.makeProgram(config.logicFiles[k])
      if (!node) {
        return undefined
      }
      return {
        in: k,
        node,
      }
    })
    .filter(nonNullable)

  const scopeContext = LogicScope.build(
    [{ node: standardLibsProgram, in: 'standard library' }].concat(
      logicPrograms
    ),
    reporter
  )

  const programNode = LogicAST.joinPrograms([
    standardLibsProgram,
    ...logicPrograms.map(x => x.node),
  ])

  const unificationContext = LogicUnify.makeUnificationContext(
    programNode,
    scopeContext,
    reporter
  )
  const substitution = LogicUnify.unify(
    unificationContext.constraints,
    reporter
  )

  const evaluationContext = LogicEvaluate.evaluate(
    programNode,
    programNode,
    scopeContext,
    unificationContext,
    substitution,
    reporter
  )

  return evaluationContext
}
