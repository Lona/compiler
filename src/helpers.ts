import { IFS } from 'buffs'
import { createModule, ModuleContext } from './logic/module'
import { Config, load } from './utils/config'
import { defaultReporter, Reporter } from './utils/reporter'

/**
 * Helpers passed to every plugins. They contain:
 *
 * - methods abstracting the file system
 * - a centralized log reporter
 * - the workspace's configuration, etc.
 */
export type Helpers = {
  fs: IFS
  reporter: Reporter
  config: Config
  module: ModuleContext
}

export default async (
  fs: IFS,
  workspacePath: string,
  options: {
    outputPath?: unknown
    reporter?: Reporter
  } = {}
): Promise<Helpers> => {
  const { reporter = defaultReporter } = options

  const helpers: Helpers = {
    fs,
    reporter,
    config: await load(fs, workspacePath),
    module: createModule(fs, workspacePath),
  }

  return helpers
}
