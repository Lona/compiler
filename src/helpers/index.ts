import { LogicAST } from '@lona/serialization'
import { config as Config } from '../utils'
import { generate as generateEvaluationContext } from './evaluation-context'
import { EvaluationContext } from './logic-evaluate'
import { createStandardLibraryResolver } from './hardcoded-mapping'
import { Reporter, defaultReporter } from './reporter'
import { FSWrapper, createFSWrapper } from './fs'
export { HardcodedMap } from './hardcoded-mapping'
export { EvaluationContext } from './logic-evaluate'
import { TraversalConfig, reduce } from './logic-traversal'
import {
  makeProgram,
  declarationPathTo,
  findParentNode,
  findNode,
} from './logic-ast'

/**
 * Helpers passed to every plugins. They contains some methods abstracting
 * the file system, the evaluation context of the workspace,
 * a reporter (to centralize the logs), the workspace's configuration, etc.
 */
export type Helpers = {
  fs: FSWrapper
  reporter: Reporter
  config: Config.Config
  /**
   * The evaluation context of the Lona Workspace.
   *
   * It contains some mapping between the identifiers and what they reference,
   * eg. `Optional.none` -> the function declaration in Prelude.logic
   * which is used to evaluate `Optional.none` to
   * `{ type: unit, memory: { type: 'unit' } }` for example. It takes care of
   * the scope of the identifier to determine the reference.
   *
   * Additionally, it keeps track of where declaration are from.
   */
  evaluationContext: EvaluationContext | undefined
  /**
   * Takes a hardcoded map `standard library -> node` and creates a function
   * which returns a node if the Logic node passed as argument is from
   * the standard library.
   *
   * Depending on the context, the hardcoded map might use the
   * evaluation context to evaluate the Logic node instead of trying to translate
   * the standard library method into the target format.
   */
  createStandardLibraryResolver: typeof createStandardLibraryResolver
  ast: {
    /** Tries to make a program out of any Logic node */
    makeProgram: (node: LogicAST.SyntaxNode) => LogicAST.Program | undefined
    /** A set of methods to help traverse the AST */
    traversal: {
      declarationPathTo: (
        node: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) => string[]
      /** Find the node with the matching id */
      findNode: (
        rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) =>
        | LogicAST.SyntaxNode
        | LogicAST.Pattern
        | LogicAST.Identifier
        | undefined
      /** Find the parent of the node with the matching id */
      findParentNode: (
        rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) => LogicAST.SyntaxNode | undefined
      /**
       * The `reduce()` method executes a reducer function (that you provide)
       * on each node of the AST, resulting in a single output value.
       *
       * Your reducer function's returned value is assigned to the accumulator,
       * whose value is remembered across each iteration throughout the AST,
       * and ultimately becomes the final, single resulting value.
       *
       * The traversal is depth-first, you can specify whether it should be
       * pre-order or post-order in the config.
       */
      reduce: <T>(
        node: LogicAST.SyntaxNode,
        callbackfn: (
          previousValue: T,
          currentNode: LogicAST.SyntaxNode,
          config: TraversalConfig
        ) => T,
        initialResult: T,
        config?: TraversalConfig
      ) => T
    }
  }
}

export type PreludeFlags = {
  [id: string]:
    | LogicAST.RecordDeclaration
    | LogicAST.EnumerationDeclaration
    | LogicAST.NamespaceDeclaration
}

export default async (
  workspacePath: string,
  outputPath?: unknown,
  _reporter?: Reporter
): Promise<Helpers> => {
  const fsWrapper = createFSWrapper(workspacePath, outputPath)

  const config = await Config.load(workspacePath, {
    forEvaluation: true,
    fs: fsWrapper,
  })

  let cachedEvaluationContext: EvaluationContext | undefined

  const reporter = _reporter || defaultReporter

  return {
    fs: fsWrapper,
    config,
    get evaluationContext() {
      if (cachedEvaluationContext) {
        return cachedEvaluationContext
      }
      cachedEvaluationContext = generateEvaluationContext(config, reporter)
      return cachedEvaluationContext
    },
    createStandardLibraryResolver,
    reporter,
    ast: {
      makeProgram,
      traversal: {
        declarationPathTo,
        findNode,
        findParentNode,
        reduce,
      },
    },
  }
}
