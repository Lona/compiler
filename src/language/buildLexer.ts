import { EnumerationDeclaration } from '../logic/nodes/EnumerationDeclaration'
import { Helpers } from '../helpers'
import { Lexer, Rule, Token, StateDefinition, Action } from './Lexer'
import { IdentifierExpression } from '../logic/nodes/IdentifierExpression'
import { LiteralExpression } from '../logic/nodes/LiteralExpression'
import { StringLiteral } from '../logic/nodes/literals'
import { LogicAST } from '@lona/serialization'
import { FunctionCallExpression } from '../logic/nodes/FunctionCallExpression'
import { isNode } from '../logic/ast'
import { valueBindingName } from '../plugins/swift/convert/codable'
import { IExpression } from '../logic/nodes/interfaces'

const getStringLiteral = (node: IExpression): string | undefined =>
  node instanceof LiteralExpression && node.literal instanceof StringLiteral
    ? node.literal.value
    : undefined

export function isTokenizer(node: EnumerationDeclaration): boolean {
  return node.attributes.some(
    attribute =>
      attribute.callee instanceof IdentifierExpression &&
      attribute.callee.name === 'tokenize'
  )
}

type TokenGenerator = {
  parentState: string
  rule: Rule
  transform: (token: Token) => unknown
}

function getAction(node: IExpression): Action {
  if (
    node instanceof FunctionCallExpression &&
    node.callee instanceof IdentifierExpression
  ) {
    switch (node.callee.name) {
      case 'pop':
        return { type: node.callee.name }
      case 'push':
      case 'next': {
        const value = getStringLiteral(node.argumentExpressionNodes[0])

        if (value) {
          return { type: node.callee.name, value }
        }
      }
    }
  }

  throw new Error('Invalid action')
}

export function getTokenAttributes(
  node: Extract<LogicAST.EnumerationCase, { type: 'enumerationCase' }>
): TokenGenerator[] {
  return node.data.attributes
    .map(attribute => new FunctionCallExpression(attribute))
    .flatMap(attribute => {
      if (
        attribute.callee instanceof IdentifierExpression &&
        attribute.callee.name === 'token'
      ) {
        const { pattern, state, action } = attribute.argumentExpressionNodes

        const rule = {
          name: node.data.name.name,
          pattern: getStringLiteral(pattern) ?? node.data.name.name,
          ...(action && { action: getAction(action) }),
        }

        const parentState = getStringLiteral(state) ?? 'main'

        const associatedValues = node.data.associatedValues.filter(isNode)

        return {
          parentState,
          rule,
          transform: token => {
            return {
              type: token.type,
              ...Object.fromEntries(
                associatedValues.map((associatedValue, index, array) => [
                  valueBindingName(associatedValue, index, array.length),
                  token.groups[index],
                ])
              ),
            }
          },
        }
      }
      return []
    })
}

export function buildLexer(
  node: EnumerationDeclaration,
  context: Helpers
): (source: string) => unknown {
  const generators = node.cases.flatMap(getTokenAttributes)

  const states: StateDefinition[] = []

  generators.forEach(generator => {
    const state = states.find(state => state.name === generator.parentState)

    if (state) {
      state.rules.push(generator.rule)
    } else {
      states.push({ name: generator.parentState, rules: [generator.rule] })
    }
  })

  const lexer = new Lexer(states, 'main')

  return source => {
    const tokens = lexer.tokenize(source)

    return tokens.map(token => {
      const generator = generators.find(
        generator => generator.rule.name === token.type
      )!
      return generator.transform(token)
    })
  }
}
