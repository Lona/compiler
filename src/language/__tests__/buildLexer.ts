import { LogicAST } from '@lona/serialization'
import { createFs } from 'buffs'
import fs from 'fs'
import path from 'path'
import { createHelpers } from '../../helpers'
import { EnumerationDeclaration } from '../../logic/nodes/EnumerationDeclaration'
import { findNode } from '../../logic/traversal'
import { buildLexer, buildTokenTransformer } from '../buildLexer'
import { formatTokens, print } from '../Printer'

it('converts SQL language', async () => {
  const source = createFs({
    'lona.json': JSON.stringify({}),
    'SQLLanguage.logic': fs.readFileSync(
      path.join(__dirname, './mocks/SQLLanguage.logic'),
      'utf8'
    ),
  })

  const helpers = createHelpers(source, '/')

  const logicFile = helpers.module.logicFiles.find(file =>
    file.sourcePath.endsWith('SQLLanguage.logic')
  )

  if (!logicFile) {
    throw new Error('Failed to find input file')
  }

  const mainEnum = findNode(
    logicFile.rootNode,
    (node): node is LogicAST.EnumerationDeclaration =>
      node.type === 'enumeration'
  )

  if (!mainEnum) {
    throw new Error('Failed to find tokens enum')
  }

  const node = new EnumerationDeclaration(mainEnum)
  const lexer = buildLexer(node, helpers)
  const transformer = buildTokenTransformer(node, helpers)

  const tokens = lexer.tokenize('SELECT foo FROM 123')

  expect(transformer(tokens)).toMatchSnapshot()
})

it('converts XML language', async () => {
  const source = createFs({
    'lona.json': JSON.stringify({}),
    'XMLLanguage.logic': fs.readFileSync(
      path.join(__dirname, './mocks/XMLLanguage.logic'),
      'utf8'
    ),
  })

  const helpers = createHelpers(source, '/')

  const logicFile = helpers.module.logicFiles.find(file =>
    file.sourcePath.endsWith('XMLLanguage.logic')
  )

  if (!logicFile) {
    throw new Error('Failed to find input file')
  }

  const mainEnum = findNode(
    logicFile.rootNode,
    (node): node is LogicAST.EnumerationDeclaration =>
      node.type === 'enumeration'
  )

  if (!mainEnum) {
    throw new Error('Failed to find tokens enum')
  }

  const node = new EnumerationDeclaration(mainEnum)
  const lexer = buildLexer(node, helpers)
  const transformer = buildTokenTransformer(node, helpers)

  const tokens = lexer.tokenize(`<hello a="test" b='foo' /><OK>Some Text</OK>`)

  expect(transformer(tokens)).toMatchSnapshot()

  const formatted = formatTokens(lexer.stateDefinitions, tokens)

  expect(formatted).toMatchSnapshot()

  expect(print(formatted)).toMatchSnapshot()
})
