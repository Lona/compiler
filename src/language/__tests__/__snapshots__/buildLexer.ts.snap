// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts SQL language 1`] = `
Array [
  Object {
    "type": "select",
  },
  Object {
    "type": "whitespace",
  },
  Object {
    "type": "identifier",
    "value": "foo",
  },
  Object {
    "type": "whitespace",
  },
  Object {
    "type": "from",
  },
  Object {
    "type": "whitespace",
  },
  Object {
    "type": "integer",
    "value": "123",
  },
]
`;

exports[`converts XML language 1`] = `
Array [
  Object {
    "type": "open",
    "value": "hello",
  },
  Object {
    "type": "name",
    "value": "a",
  },
  Object {
    "type": "equals",
  },
  Object {
    "type": "string",
    "value": "test",
  },
  Object {
    "type": "name",
    "value": "b",
  },
  Object {
    "type": "equals",
  },
  Object {
    "type": "string",
    "value": "foo",
  },
  Object {
    "type": "slashClose",
  },
  Object {
    "type": "open",
    "value": "OK",
  },
  Object {
    "type": "close",
  },
  Object {
    "type": "text",
    "value": "Some Text",
  },
  Object {
    "type": "slashOpen",
  },
  Object {
    "type": "name",
    "value": "OK",
  },
  Object {
    "type": "close",
  },
]
`;

exports[`converts XML language 2`] = `
Object {
  "parts": Array [
    Object {
      "parts": Array [
        "<",
        "hello",
      ],
      "type": "concat",
    },
    "a",
    "=",
    Object {
      "parts": Array [
        "\\"",
        "test",
        "\\"",
      ],
      "type": "concat",
    },
    "b",
    "=",
    Object {
      "parts": Array [
        "\\"",
        "foo",
        "\\"",
      ],
      "type": "concat",
    },
    "/>",
    Object {
      "parts": Array [
        "<",
        "OK",
      ],
      "type": "concat",
    },
    ">",
    "Some Text",
    "</",
    "OK",
    ">",
  ],
  "type": "concat",
}
`;

exports[`converts XML language 3`] = `"<helloa=\\"test\\"b=\\"foo\\"/><OK>Some Text</OK>"`;
